### ISSUE TITLE
Scrollbar bleeding through header bar

### ISSUE THREAD

**Issue by polarathene** [squidfunk/mkdocs-material#2428]
- [x] I've read the [contribution guidelines][1] and agree with them

__I've found a bug and checked that ...__

- [x] ... the problem doesn't occur with the default MkDocs template
- [x] ... the problem is not in any of my customizations (CSS, JS, template)
- [x] ... the documentation does not mention anything about my problem
- [x] ... there are no open or closed issues that are related to my problem

## Description

Scrollbars are bleeding through the header bar. This may be intentional for a vertical scrollbar (which doesn't pad it's height against the header bar), but unlikely desirable for horizontal scrollbars.

An easy fix, but not fully tested is adjusting the z-index for `.md-content` to be lower than the header bar sibling element (has z-index of 2).

### Expected behavior

![Screenshot_20210311_121727](https://user-images.githubusercontent.com/5098581/110711052-c4cbdd00-8263-11eb-97af-d88aecd1b791.png)

### Actual behavior

![Screenshot_20210311_115444](https://user-images.githubusercontent.com/5098581/110710931-87674f80-8263-11eb-9a81-7071dad3f577.png)

### Steps to reproduce the bug

1. Visit https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#adding-annotations
2. View at width <540 px
3. Find code block with horizontal scrollbar
4. Scroll down until it bleeds through headbar at top of page.

### Package versions / Project configuration / System information

Active release used for project demo site: https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#adding-annotations

**Comment by polarathene** [comment]
- Screenshots from Firefox 85 (Manjaro KDE)
- Chrome Android, unable to reproduce.
- Firefox Android, Firefox Focus (Android, different webengine), both hide the scrollbar visibility until touch interaction. Although I could scroll vertically while having it visible, and could not reproduce.

May be Linux specific if the scrollbar is GTK based. All I know is it bleeds through and z-index resolves it.

**Comment by squidfunk** [comment]
Thanks for reporting. Does it happen on the [Insiders docs](https://squidfunk.github.io/mkdocs-material-insiders/)? It has slightly different indexes (had to adjust them for some features), so maybe it's already fixed there.

**Comment by polarathene** [comment]
> Does it happen on the [Insiders docs](https://squidfunk.github.io/mkdocs-material-insiders/)?

Yes, problem exists there too.

**Comment by squidfunk** [comment]
Thanks. I'm not able to fix this, as it seems to be local to Linux/GTK. PR appreciated.

**Comment by squidfunk** [comment]
On a side note, even though we can fix it, I’d expect this to break regularly, because I cant test on this systemas most of the changes on this project are done by me. Basically the same as with Japanese / Chinese search, which breaks rather often. Just too may combinations I guess 🤷‍♀️

**Comment by polarathene** [comment]
It's fixable by setting `md-content` z-index to be lower than the header bar z-index. As mentioned above. They're sibling elements, so I don't know how it's intended to break regularly unless you have messy hacks that would break z-order and not show up on other browsers.

I imagine the difference in behaviour is because z-index is not explicit, so setting it for siblings would do the opposite of breakage (assuming a sane order remains), vs implicit ordering.

You can technically test this AFAIK, via CI, but I understand that the effort of putting a test case together for it isn't a good use of your time.

Japanese/Chinese search, I've not looked into those issues but they could be due to unicode issues. A good example of where that goes wrong is emoji input and how fragile that has been found to be. My own experience when porting a hashing algorithm to JS was parsing some unicode characters, the naive implementation I had only read a portion of the actual character and thus resulted in a hash output that differed from implementations in other programming languages which handle the content of strings differently (among other JS specific issues I had to account for).

**Comment by squidfunk** [comment]
So to verify, setting `z-index: 0` fixes it? It would be lower than the one of the header, and `0` should be safe to not break anything.

**Comment by polarathene** [comment]
> So to verify, setting `z-index: 0` fixes it?

Yes, when setting `.md-content` (or sibling element to the `.md-header`: `.md-container`) to include a `z-index: 0` style the issue is resolved.

---

Even with `z-index: 1` it's still resolved fine as `.md-header` is using `z-index: 2`. If matching to `z-index: 2` `.md-container` layer overlays the `.md-header`. Any `z-index` value of children like `.md-content` that would have an effect without `.md-container` setting an explicit `z-index` value also seem to be ignored and `.md-container` keeps it's `z-index` to `.md-header` intact.

`.md-typeset pre` for a code block sets `position: relative;` which allows `z-index` to have an effect (provided no parent element z-index takes precedence for upper layers), otherwise `.md-content` seems to be the lowest in that hierarchy to have any impact with `z-index` here.

**Comment by squidfunk** [comment]
I'm trying to understand how many users are impacted by this minor rendering bug:

> - Screenshots from Firefox 85 (Manjaro KDE)
> - Chrome Android, unable to reproduce.
> - Firefox Android, Firefox Focus (Android, different webengine), both hide the scrollbar visibility until touch interaction. Although I could scroll vertically while having it visible, and could not reproduce.

I read that it's only reproducible on Firefox, Linux? In that case, I'd be more comfortable leaving the `z-index` as is, because introducing this might have unintended side effects with combinations of other features. The reasoning behind this is explained in https://github.com/squidfunk/mkdocs-material/issues/2432#issuecomment-798077053.

**Comment by polarathene** [comment]
I don't personally see anything wrong with having explicit z-index ordering rather than mixing implicit with explicit as is currently done. I doubt it would introduce more problems when you're assisting the browser to better understand the intended ordering.

It's fine, I understand that it's likely a rendering bug with Firefox on Linux that needs to be reported there for a fix instead. Cheers.

**Comment by squidfunk** [comment]
> I don't personally see anything wrong with having explicit z-index ordering rather than mixing implicit with explicit as is currently done. I doubt it would introduce more problems when you're assisting the browser to better understand the intended ordering.

The thing is that `z-index` only has an effect for elements that have a `position` different from `static`, which is not the case for `.md-content`. It's definitely a Firefox bug, as it shouldn't have any effect. Adding that _shouldn't_ introduce more problems, but it might.

**Comment by polarathene** [comment]
> The thing is that `z-index` only has an effect for elements that have a `position` different from `static`, which is not the case for `.md-content`.

I understand that, but if the `z-index`  is higher than `.md-header` that entire layer sits above the header content, so text would overlay the header too given a higher `z-index` on `.md-content`.

Firefox does indicate/detect when applying `z-index` to an element would have no effect, and points out that it's due to the `position` not being different from `static` as you mention. So it must be detecting something that affects `.md-content` in that way for `z-index` to be valid when it's otherwise not for everything below? (**EDIT:** `z-index` is valid on the actual code block that sets it's own `position: relative`)

Perhaps it's a bug on Firefox, that's fine, although I can understand how Firefox is probably revealing the issue because it might be confused about what the implicit `z-index` for other elements should be? (seems to be correct for everything else in the `.md-content` layer except the scrollbar which isn't respecting the implicit `z-index` of `.md-content` for whatever reason, presumably because it's a more native UI element with it's own implicit `z-index`?)

---

**EDIT:** I found out that there are various conditions that [trigger `z-index` validity](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/The_stacking_context), one being a child of a flex parent, which applies to all siblings here as the parent `<body>` is a flex container.

<details>
<summary>Additional Insights</summary>

Additionally from that MDN link:

> Within a stacking context, child elements are stacked according to the same rules previously explained. Importantly, the z-index values of its child stacking contexts only have meaning in this parent. Stacking contexts are treated atomically as a single unit in the parent stacking context.

and:

> Each stacking context is completely independent of its siblings: only descendant elements are considered when stacking is processed.
> Each stacking context is self-contained: after the element's contents are stacked, the whole element is considered in the stacking order of the parent stacking context.
> **We can say that elements that do not create their own stacking contexts are assimilated by the parent stacking context.**

I think in this case, the scroll bar UI is intended to have a natural high `z-index` for UX reasons by default, thus why it bleeds through in this case, because it's not been given an explicit `z-index` from the sibling, so no overriding or self-containment within `<body>`?

I've not verified in other browsers but setting `.md-typeset pre` to have `z-index: 100` will have the full code block bleed through as the MDN docs would suggest. I have then set `.md-container` to have an explicit `z-index: 1` and this as expected by spec, placed the container layer below the header, the code block z-index stays within the explicit stacking context and doesn't bleed through (or get assimilated into parent stacking context as it mentions).

We can see also from [this MDN doc](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/Adding_z-index):

> When no z-index property is specified, elements are rendered on the default rendering layer 0 (zero).
> If several elements share the same z-index value (i.e., they are placed on the same layer), stacking rules explained in the section [Stacking without the z-index property](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/Stacking_without_z-index) apply.

The referenced link in the quote to another MDN page on the subject states:

> When the [`z-index`](https://developer.mozilla.org/en-US/docs/Web/CSS/z-index) property is not specified on any element, elements are stacked in the following order (from bottom to top):
> 
> 1.  The background and borders of the root element
> 2.  Descendant [non-positioned](https://developer.mozilla.org/en-US/docs/Web/CSS/position#static) blocks, in order of appearance in the HTML
> 3.  Descendant [positioned](https://developer.mozilla.org/en-US/docs/Web/CSS/position#types_of_positioning) elements, in order of appearance in the HTML
>
> Keep in mind, when the [`order`](https://developer.mozilla.org/en-US/docs/Web/CSS/order) property alters rendering from the "order of appearance in the HTML" within [`flex`](https://developer.mozilla.org/en-US/docs/Web/CSS/flex) containers, it similarly affects the order for stacking context.

I know that you've got some usage of `order`, along with multiple elements in the same stacking context sharing the same `z-index` value, the children of `<body>` has several with `z-index: 2` with `.md-container` inbetween.

With all these rules, it's no wonder why browsers might get confused in implementation and render differently, but it would seem like while Firefox on Linux is a niche case, it's technically respecting the spec/rules, and that at any point in future, the other browsers could be affected. In this case, for whatever reason it'd appear that scrollbars just have higher (or explicit) z-index natively on Firefox Linux I guess?

---

</details>

**The correct solution is to be properly explicit about your `z-index` when adjusting the stacking order**, rather than relying on implicit behaviour, implementation bugs or differences that are more prone to change.

**Setting `.md-typeset pre` to include `z-index: 0` also resolves the issue.**

It seems a [similar issue affected Chrome on macOS a while back](https://stackoverflow.com/questions/16874546/strange-z-index-behavior-with-scrollbars-under-chrome) (last user chiming in about it on that issue was late 2017), where the solution was to use something that creates a stacking context to resolve the issue. 

[Another user reporting the same issue on Firefox (OS unclear) in 2020](https://stackoverflow.com/questions/59933191/how-to-prevent-firefox-scrollbars-from-overlapping-all-content). I don't have Windows or macOS on hand, from what I can see on Android browsers, scrollbars may take priority

**Comment by polarathene** [comment]
Well... this is embarrassing.. I was looking through various reported bugs similar this one and I came across one from [less than a year ago](https://bugzilla.mozilla.org/show_bug.cgi?id=1641567). It gets a response that the "bug" is due to Firefox dev tools [Responsive Design Mode](https://developer.mozilla.org/en-US/docs/Tools/Responsive_Design_Mode) (RDM) is applying some additional stylesheet overrides.

So the scrollbar priority is due to that, although correctly managing the `z-index` does resolve this issue.

Windows and macOS are probably affected in the same manner...but nobody is going to be viewing the docs for the content while having the dev tools up :sweat_smile: Closing those and resizing the window manually to get the scrollbars showing, no rendering issue, whoops! :grimacing:
